src/backend/dx12/src/device.rs

```diff
@@ -266,20 +266,133 @@ impl d::Device<B> for Device {
         dependencies: &[pass::SubpassDependency],
     ) -> n::RenderPass {
-        // TODO:
-        let subpasses = subpasses
+        #[derive(Copy, Clone, Debug, PartialEq)]
+        pub enum SubState {
+            New(winapi::D3D12_RESOURCE_STATES),
+            Preserve,
+            Undefined,
+        }
+        struct AttachmentInfo {
+            sub_states: Vec<SubState>,
+            target_state: winapi::D3D12_RESOURCE_STATES,
+            last_state: winapi::D3D12_RESOURCE_STATES,
+            barrier_start_index: usize,
+        }
+
+        let mut att_infos = attachments
             .iter()
-            .map(|subpass| {
-                n::SubpassDesc {
-                    color_attachments: subpass.color_attachments.iter().cloned().collect(),
+            .map(|att| AttachmentInfo {
+                sub_states: vec![SubState::Undefined; subpasses.len()],
+                target_state: if att.format.0.is_depth() {
+                    winapi::D3D12_RESOURCE_STATE_DEPTH_WRITE //TODO?
+                } else {
+                    winapi::D3D12_RESOURCE_STATE_RENDER_TARGET
+                },
+                last_state: conv::map_image_resource_state(image::Access::empty(), att.layouts.start),
+                barrier_start_index: 0,
+            })
+            .collect::<Vec<_>>();
+
+        // Fill out subpass known layouts
+        for (sid, sub) in subpasses.iter().enumerate() {
+            for &(id, _layout) in sub.color_attachments {
+                let state = SubState::New(att_infos[id].target_state);
+                let old = mem::replace(&mut att_infos[id].sub_states[sid], state);
+                debug_assert_eq!(SubState::Undefined, old);
+            }
+            for &(id, _layout) in sub.input_attachments {
+                let state = SubState::New(winapi::D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE);
+                let old = mem::replace(&mut att_infos[id].sub_states[sid], state);
+                debug_assert_eq!(SubState::Undefined, old);
+            }
+            for &id in sub.preserved_attachments {
+                let old = mem::replace(&mut att_infos[id].sub_states[sid], SubState::Preserve);
+                debug_assert_eq!(SubState::Undefined, old);
+            }
+        }
+
+        let mut deps_left = vec![0u16; subpasses.len()];
+        for dep in dependencies {
+            if dep.passes.start != pass::SubpassRef::External {
+                if let pass::SubpassRef::Pass(sid) = dep.passes.end {
+                    deps_left[sid] += 1;
                 }
-            }).collect();
+            }
+        }
 
-        n::RenderPass {
+        let mut rp = n::RenderPass {
             attachments: attachments.to_vec(),
-            subpasses,
-            dependencies: dependencies.to_vec(),
+            subpasses: Vec::new(),
+            post_barriers: Vec::new(),
+        };
+
+        while let Some(sid) = deps_left.iter().position(|count| *count == 0) {
```

Comment:
Intuitively, I would say this should be `*count != 0` because we check above for which subpasses we need to insert dependencies, shouldn't it?

Reply:
The array stores the number of dependencies that this subpass needs to be fulfilled before it can start. So `count==0` would mean we are good to go, after which we scan over all the dependencies that start off this subpass and deduce them from the descination subpasses.